elastic/kibana
# [Upgrade Assistant] Show reindex warning if low disk space

## Summary

Related to https://github.com/elastic/kibana/issues/123040.

This PR adds a warning callout to the reindexing flyout in the event that one or more of the user's nodes has exceeded the [low disk watermark setting](elastic.co/guide/en/elasticsearch/reference/current/modules-cluster.html#disk-based-shard-allocation).

### How is this calculated?

We can access the low disk watermark setting via the [ES cluster settings API](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-get-settings.html). Then, we can compare this to the available and total disk space on the file system using the [ES node stats API](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-stats.html).

### How to test

There are a few different scenarios to test. All require you have a 7.17 cluster running pointing to 6.x data with old indices.

1. When the low disk watermark setting has not been exceeded.
    - UA should be have normal. No deprecation should appear on the ES deprecations page, or the reindexing flyout.

2. When the low disk watermark setting is a percentage value, and the percentage disk used exceeds this value.
   - Navigate to Console, and set a very low disk watermark setting using a percentage value. For example:
   
```
  PUT /_cluster/settings
    {
        "transient": {
            "cluster.routing.allocation.disk.watermark.low" : "50%"
        }
    }
```

Verify deprecation displays in list. Verify warning callout displays in the reindex flyout.

3. When the low disk watermark setting is a bytes value, and the setting is less than the disk space available.
   - Navigate to Console, and set a low disk watermark setting using a bytes value. Note `flood_stage` and `high` watermark values will also need to be converted to bytes. For example:
  
```
  PUT /_cluster/settings
  {
      "transient": {
          "cluster.routing.allocation.disk.watermark.low" : "2000gb",
          "cluster.routing.allocation.disk.watermark.flood_stage": "100mb",
          "cluster.routing.allocation.disk.watermark.high": "100mb"
      }
  }
```

Verify deprecation displays in list. Verify warning callout displays in the reindex flyout.

### Screenshots

![Upgrade-Assistant-Elastic](https://user-images.githubusercontent.com/5226211/155722441-a978af0f-d806-4ffd-a8f2-d547a50ac139.png)
